File: NodeAuth/src/controllers/registerController.ts
import { Router, type Request, type Response } from "express";

import type { AppConfig } from "../config";
import type { ConfirmEmailRequest } from "../models/ConfirmEmailRequest";
import type { RegisterRequest } from "../models/RegisterRequest";
import type { RegistrationEmailRequest } from "../models/RegistrationEmailRequest";
import { RegistrationService } from "../services/RegistrationService";
import { validateOrigin } from "../utils/originValidator";

type RegisterControllerDeps = {
  registrationService: RegistrationService;
  config: AppConfig;
};

export const buildRegisterRouter = ({
  registrationService,
  config,
}: RegisterControllerDeps): Router => {
  const router = Router();

  router.post(
    "/web-register",
    async (req: Request, res: Response): Promise<void> => {
      const originResult = validateOrigin(req, config.origin.host);
      if (!originResult.allowed) {
        console.warn(originResult.error);
        res.sendStatus(403);
        return;
      }

      const body = req.body as RegisterRequest;
      if (!body.email || !body.password) {
        res.sendStatus(400);
        return;
      }

      const result = await registrationService.registerAsync(
        body.email,
        body.password,
      );
      if (!result.success) {
        res.status(400).json({
          error: result.error,
          error_description: result.errorDescription,
        });
        return;
      }

      res.status(200).json({
        message:
          "Registration successful. Please check your email to confirm your account.",
      });
    },
  );

  router.post(
    "/confirm-email",
    async (req: Request, res: Response): Promise<void> => {
      const originResult = validateOrigin(req, config.origin.host);
      if (!originResult.allowed) {
        console.warn(originResult.error);
        res.sendStatus(403);
        return;
      }

      const body = req.body as ConfirmEmailRequest;
      if (!body?.token) {
        res.status(400).json({
          error: "TokenRequired",
          error_description: "Confirmation token is required.",
        });
        return;
      }

      const result = await registrationService.confirmEmailAsync(body.token);
      if (!result.success) {
        res.status(400).json({
          error: result.error,
          error_description: result.errorDescription,
        });
        return;
      }

      res.status(200).json({
        message: "Email confirmed successfully.",
      });
    },
  );

  router.post(
    "/send-registration-email",
    async (req: Request, res: Response): Promise<void> => {
      const originResult = validateOrigin(req, config.origin.host);
      if (!originResult.allowed) {
        console.warn(originResult.error);
        res.sendStatus(403);
        return;
      }

      const body = req.body as RegistrationEmailRequest;
      if (!body?.email) {
        res.status(400).json({
          error: "EmailRequired",
          error_description: "Email is required.",
        });
        return;
      }

      await registrationService.resendRegistrationEmail(body.email);
      res.sendStatus(200);
    },
  );

  return router;
};

File: Auth/Auth.Web/Controllers/RegisterController.cs
using Auth.App;
using Auth.Web.Configuration;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;
using System.ComponentModel.DataAnnotations;
using Users.App;

namespace Auth.Web.Controllers
{
    [ApiController]
    [Route(GlobalConstants.BasePath)]
    public class RegisterController(RegistrationService registration, OriginValidator originValidator, ILogger<RegisterController> logger)
        : ControllerBase
    {
        [HttpPost("web-register")]
        [ResponseCache(NoStore = true, Location = ResponseCacheLocation.None)]
        public async Task<ActionResult> Register([FromBody] RegisterRequestDto req)
        {
            if (!IsAllowedOrigin(Request))
            {
                logger.LogWarning(originValidator.Error);
                return Forbid();  // simple CSRF guard 
            }

            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var result = await registration.RegisterAsync(req.Email, req.Password);
            if (!result.Success)
            {
                if (result.Error == "MaxConfirmationEmailsPerDay")
                {
                    return BadRequest(new { error = result.Error, error_description = result.ErrorDescription });
                }
                return BadRequest(new { error = result.Error, error_description = result.ErrorDescription });
            }

            return Ok(new { message = "Registration successful. Please check your email to confirm your account." });
        }

        [HttpPost("confirm-email")]
        [ResponseCache(NoStore = true, Location = ResponseCacheLocation.None)]
        public async Task<ActionResult> ConfirmEmail([FromBody] ConfirmEmailRequestDto req)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(new { error = "TokenRequired", error_description = ValidationConstants.ConfirmEmailTokenRequired });
            }

            var result = await registration.ConfirmEmailAsync(req.Token);
            if (!result.Success)
            {
                return BadRequest(new { error = result.Error, error_description = result.ErrorDescription });
            }

            return Ok(new { message = "Email confirmed successfully." });
        }

        [HttpPost("send-registration-email")]
        public async Task<IActionResult> SendRegistrationEmail([FromBody] RegistrationEmailRequestDto req)
        {
            await registration.ResendRegistrationEmail(req.Email);
            return Ok();
        }

        private bool IsAllowedOrigin(HttpRequest r) => originValidator.IsAllowedOrigin(r);
    }

    public record RegisterRequestDto(
        [Required(ErrorMessage = "Email is required.")]
        [EmailAddress(ErrorMessage = "Invalid email format.")]
        string Email,
        [Required(AllowEmptyStrings = false, ErrorMessage = ValidationConstants.PasswordRequired)]
        [RegularExpression(ValidationConstants.PasswordPattern, ErrorMessage = ValidationConstants.PasswordFormatError)]
        string Password
    );

    public record ConfirmEmailRequestDto(
        [Required(AllowEmptyStrings = false, ErrorMessage = ValidationConstants.ConfirmEmailTokenRequired)]
        string Token
    );

    public record RegistrationEmailRequestDto(
        [Required(ErrorMessage = ValidationConstants.EmailRequired)]
        [RegularExpression(ValidationConstants.EmailPattern, ErrorMessage = ValidationConstants.EmailFormatError)]
        string Email
    );
}
